axlearn/open_api/metrics/tool_use_execution_utils.py

```diff
@@ -0,0 +1,307 @@
+    words = normalized_str_value.split()
+    while len(words) > 1 and words[0] in _STOP_WORDS:
+        words = words[1:]
+    while len(words) > 1 and words[-1] in _STOP_WORDS:
+        words = words[:-1]
+    normalized_str_value = " ".join(words)
+    return normalized_str_value
+
+
+def _word_set(input_str: str) -> set[str]:
+    """Returns a set of words splitting by whitespace, newline and tab,
+    and skipping empty strings from the set."""
+    return {s for s in re.split(r"\s+", input_str) if s}
+
+
+def _match_strings_bag_of_words(pred_str: str, target_str: str, threshold: float = 1.0) -> bool:
+    """Match strings using a bag of words approach.
+
+    Args:
+        pred_str: The predicted argument string.
+        target_str: The target argument string.
+        threshold: applied to (# unique common words) / (# unique pred_str words)
+
+    Returns:
+        True if recall is higher or equal to threshold; otherwise return False.
+    """
+    if pred_str == target_str:
+        return True
+
+    assert threshold > 0, "bag of words string matching threshold must be above 0"
+    pred_word_set = _word_set(pred_str)
+    target_word_set = _word_set(target_str)
+    common_words = target_word_set.intersection(pred_word_set)
+    if len(pred_word_set) == 0:
+        return False
+    ratio = len(common_words) / len(target_word_set)
+    return ratio >= threshold
+
+
+def _is_arg_value_equal(
+    pred_arg: ValueOrListOf,
+    target_arg: ValueOrListOf,
+    check_lenient: bool,
+    bag_of_words: bool,
+) -> bool:
+    """Checks if the predicted and target arguments are equal under different checks."""
+    if check_lenient:
+        if (
+            isinstance(pred_arg, list)
+            and isinstance(target_arg, list)
+            and len(pred_arg) == len(target_arg)
+        ):
+            return all(
+                _is_arg_value_equal(
+                    pred_arg=el_pred,
+                    target_arg=el_target,
+                    check_lenient=check_lenient,
+                    bag_of_words=bag_of_words,
+                )
+                for el_pred, el_target in zip(pred_arg, target_arg)
+            )
+        # Only handling string payloads for lenient evaluation
+        if isinstance(pred_arg, str) and isinstance(target_arg, str):
+            pred_lenient = _string_lenient_transform(pred_arg)
+            target_lenient = _string_lenient_transform(target_arg)
+            if bag_of_words:
+                return _match_strings_bag_of_words(
+                    pred_str=pred_lenient, target_str=target_lenient, threshold=1.0
+                )
+            return pred_lenient == target_lenient
+    return pred_arg == target_arg
+
+
+def check_arguments(
+    pred_args: Dict[str, ValueOrListOf],
+    target_args: Dict[str, ValueOrListOf],
+    check_lenient: bool = False,
+    bag_of_words: bool = False,
+) -> bool:
```

Comment:
```suggestion
def check_arguments(
    *,
    pred_args: dict[str, ValueOrListOf],
    target_args: dict[str, ValueOrListOf],
    check_lenient: bool = False,
    bag_of_words: bool = False,
) -> bool:
```

FWIW, this kind of API with multiple bool flags is generally discouraged: https://docs.google.com/document/u/1/d/e/2PACX-1vQGhAmTlrw0WXSnhdGsFjafjq2ZaxTtUsXq7LOxHkZf5Zr5MO5UQIcF5sXev5uBjDFHJRW0p6txExb8/pub#h.4mp6cu6147ou

E.g., specifying `bag_of_words=True` actually has no effect unless `check_lenient=True` is also specified.

Reply:
Good suggestion. Done